### DISCUSSION TITLE
Polycrystalline simulation with crystal plasticity

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#16716] [Q&A Modules: Solid mechanics] [answered]
Hello all,

I am trying to build and run a polycrystalline crystal plasticity model, which uses Voronoi tesslation to establish the initial crystal structure. But there are no examples of polycrystalline modeling in the folder of cp_uer_object or other crystal plasticity examples. 

So what I want to ask is, would you share me the  input file related to it?

Any help would be greatly appreciated!

Thanks,

Wei Peng

**Comment by jiangwen84** [comment]
Two options: 

1. assign each grain with a block id, and use `read_type = block` in the following block. The input file can be found at modules/tensor_mechanics/test/tests/cp_user_object/prop_block_read.i

[UserObjects]
  [./prop_read]
    type = ElementPropertyReadFile
    prop_file_name = 'euler_ang_file.txt'
    # Enter file data as prop#1, prop#2, .., prop#nprop
    nprop = 3
    read_type = block
    nblock= 2
  [../]
[]

2. use `ElementPropertyReadFile` to perform voronoi tessellation. 

[UserObjects]
  [./prop_read]
    type = ElementPropertyReadFile
    prop_file_name = 'euler_ang_file.txt'
    # Enter file data as prop#1, prop#2, .., prop#nprop
    nprop = 3
    read_type = grain
    ngrain= 2
  [../]
[]

**Reply by PengWei97** [reply]
Dear @jiangwen84,

Thanks for your reply. I have understood the first option you mentioned. But for the second option, I have a question that needs your answer. If I use voronoi tessellation to generate the initial polycrystalline structure, is the object used is the PolycrystalVoronoi built in moose, or do I need to import the initial polycrystalline structure from an external file?

Finally, could you please provide the input file in this regard for me to solve this problem, thank you very much to jiangwen84 for your help.

Kind regards,

Wei

**Reply by jiangwen84** [reply]
For the second approach, you need  a mesh generator software, such as Cubit, Gmsh, Neper etc, to generate a mesh of polycrystalline. Each grain will be assigned a block id. Once you make such mesh, you can use `FileMeshGenerator` to read in the mesh, and use the  `ElementPropertyReadFile` to assign Euler angles to each block. 

You might want to start with simple case,  i.e. two blocks, and get familiar with the usage.

**Comment by ngrilli** [comment, marked as answer]
Dear @PengWei97 

I use neper to generate the polycrystal mesh and then I read the mesh file .msh with MOOSE.
You can find an example with instructions in my repository:
https://github.com/ngrilli/c_pfor_am/tree/main/test/tests/polycrystal

Best Regards,
Nicolò Grilli
National University of Singapore

**Reply by PengWei97** [reply]
Hi @ngrilli ,

Thank you very much for your response. I see polycrystal_out.e, but it is not a polycrystalline structure produced by voronoi tessellation. Maybe my question is not clear. What I want is to generate a polycrystalline model similar to the figure below to simulate crystal plasticity. If you could provide relevant documents in this regard, I really appreciate it..
<img width="594" alt="Snipaste_2021-01-20_11-40-10" src="https://user-images.githubusercontent.com/56183775/105124082-521c6c00-5b14-11eb-9a51-24ea0ea8f2b2.png">

**Reply by ngrilli** [reply]
Dear @PengWei97 

I think my file was produced using the "grain growth model" available in neper.
Voronoi is the default option in neper.

For instance, the following set of commands will produce a Voronoi tessellation:

neper -T -n 2 -id 1 -domain "cube(3,3,3)"
neper -M n2-id1.tess -elttype 'hex' -cl 1.0

while the following set of commands produce a grain growth tessellation:

neper -T -n 2 -id 1 -domain "cube(3,3,3)" -morpho 'gg'
neper -M n2-id1.tess -elttype 'hex' -cl 1.0

Finally, grain growth tessellation is usually preferred to Voronoi,
because it produces more realistic microstructures.

Best Regards,
Nicolò Grilli
National University of Singapore

**Reply by ngrilli** [reply]
Please check the neper manual for all tessellation options

**Reply by PengWei97** [reply]
Thank you very much for your careful guidance. I will read them carefully.

**Comment by p-antonioni** [comment]
Hello, everyone.

I'm trying to simulate the plasticity of polycrystals. For this reason, I used Neper to construct a voronoi tasselation, which I then imported using FileMeshGenerator. Because I need to incorporate Euler angles, so I created a UserObject that read attributes from a  .csv file using voronoi read_type. Then I utilized the object within Compute Elasticity Tensor CP.
The csv file has six columns: three for voronoi tessellation positions (use_random_voronoi = false) and three for Euler angles.
The issue is that the tessellation positions are set as Euler angles.
Am I doing something wrong?
I saw issue #16716 and was wondering if it was feasible to complete this type of operation using simply PropertyReadFile.
```
[UserObjects]
  [red_prop]
    type = PropertyReadFile #Running says to use PropertyReadFile instead of elementPropertyReadFile
    prop_file_name = 'threecry.csv'
    use_random_voronoi = false
    read_type = 'voronoi'
    nvoronoi = 3
    nprop = 6
  []
[]

[Materials]
  [elasticity_tensor]
    type = ComputeElasticityTensorCP
    C_ijkl = '1.684e5 1.214e5 1.214e5 1.684e5 1.214e5 1.684e5 0.754e5 0.754e5 0.754e5'
    fill_method = symmetric9
    read_prop_user_object = red_prop
  []
```
Thanks,
Paolo Antonioni

**Reply by GiudGiud** [reply]
Hello

I think the thing to do here is simply to ouput the material properties (there s a parameter on every material object for that) and check that the distributions are as expected.

Guillaume

**Reply by p-antonioni** [reply]
Thanks Guillaume for your answer. 
Do you mean to use the output_properties parameter in Compute Elasticity Tensor CP ?  
If yes, is there a way to extract all of them without providing a list?
For a single crystal simulation I provided the euler angles directly using euler_angle_ number parameters.
For a polycrystal, I provided a file with x,y,z,euler_angle1,euler_angle2,euler_angle3.
I saw from Paraview that the value of x was assigned to euler_angle_1 and similarly for other grains as well.
Did I miss some steps? Do I need to create some further material object?

Thanks, 
Paolo

**Reply by GiudGiud** [reply]
> If yes, is there a way to extract all of them without providing a list?

not sure. Does setting outputs = 'name of the output exodus' suffice?

**Reply by GiudGiud** [reply]
> I saw from Paraview that the value of x was assigned to euler_angle_1 and similarly for other grains as well.

So the property read file is accessing the wrong column?
We'd have to check the implementation of ComputeElasticityTensorCP

For now I would suggest shuffling columns around to just make it work.

**Reply by p-antonioni** [reply]
I was not able to use it, I think I am making mistake.
I have the following error message:
unused parameter 'Materials/elasticity_tensor/output'
I have also defined updated euler angles in AuxVariables and Postsprocessors 
If you prefer, I can share my input file.

**Reply by GiudGiud** [reply]
`outputs`
you can check the object parameters there:
https://mooseframework.inl.gov/source/materials/crystal_plasticity/ComputeElasticityTensorCP.html

**Reply by p-antonioni** [reply]
Thanks for your help and patience.
I am rather new to MOOSE and I am not getting how to use outputs and output_properties to output the material properties in this case.
Could you suggest me how to modify the input file to be able to do this:
[polycrystal.txt](https://github.com/idaholab/moose/files/14130271/polycrystal.txt)
The x,y,z coordinates that i use in the csv file are what neper calls seed coordinates.

**Reply by GiudGiud** [reply]
Use Debug/show_material_props=true to see the names of the material properties.

then in the material you can have `output_material_properties = ....` (the names from the debug log)

**Reply by p-antonioni** [reply]
Thanks again.
It appears that the x, y, and z coordinates in the csv file are utilized as both Voronoi tessellation coordinates and Euler angle values.
I've included a picture.
In this scenario, the tessellation appears to have been successfully recreated, but the euler_angles_1 values are the three grains' x-coordinates.
![euler_angles](https://github.com/idaholab/moose/assets/154326939/3e0aa20a-effa-4d91-8918-79158256be52)

**Reply by GiudGiud** [reply]
It will take a little modification in the material to access the right column then. 

I can look at it Monday i m on travel right now. Maybe someone else on the forum can check this though

**Reply by p-antonioni** [reply]
Hi,
Sorry for the delay in replying and thank you again. Were you able to check it?
Thanks

**Reply by GiudGiud** [reply]
Hello

No I have not gotten to it.
I have a paper to finish early next week so maybe next week.

Yiou should be able to just add dummy columns or remove one to get the data loaded the right way. And if not, adding an offset in the PropertyReadFile then re-compiling should not take long either

Guillaume

**Reply by p-antonioni** [reply]
Hi Guillaume,

Issue solved. Thanks for your help and guidance.

Paolo

**Reply by GiudGiud** [reply]
Sorry for the delay. Still dealing with papers.
I d need your mesh and your threecry.csv file to check why it s not being read as you expect.

**Reply by p-antonioni** [reply]
Hi Guillaume, 
here are the .csv file and msh:
[threecry.csv](https://github.com/idaholab/moose/files/14540736/threecry.csv)
[n3-id1.txt](https://github.com/idaholab/moose/files/14540745/n3-id1.txt) (it is a .msh file generated with neper)

If I understand properly, PropertyReadFile accesses the first three components of the csv file to find the centers, but ComputeElasticityTensorCP also utilizes the first three elements to assign Euler angles.

```
PropertyReadFile   
// Read tesselation from file
  else
  {
    for (const auto i : make_range(_nvoronoi))
      for (const auto j : make_range(_mesh.dimension()))
        _center[i](j) = _reader.getData(i)[j];
  }

ComputeElasticityTensorCP
    _Euler_angles_mat_prop[_qp](0) = _read_prop_user_object->getData(_current_elem, 0);
    _Euler_angles_mat_prop[_qp](1) = _read_prop_user_object->getData(_current_elem, 1);
    _Euler_angles_mat_prop[_qp](2) = _read_prop_user_object->getData(_current_elem, 2);

```

To solve it quickly. I just set an offset in ComputeElasticityTensorCP and the new indices are 3,4,5. I used the voronoi read type, so I have x, y, and z coordinates for the voronoi centers and three Euler angles.

Now I am trying to modify  PropertyReadFile to insert an offset and reverse the order of euler angles and centers in csv file.

**Reply by GiudGiud** [reply]
thanks! I think the APIs are not set up quite the right way if this is so.
At least they should give you warnings when this is read improperly

**Reply by GiudGiud** [reply]
it turns out I m still missing the `input_slip_sys.txt`

**Reply by GiudGiud** [reply]
ok never mind I made a dummy one

**Reply by GiudGiud** [reply]
You understood this correctly. I think we need to make this more obvious in the reader user objects so downstream objects dont need to plan for an offset in the columns when using the voronoi mode. 
I ll make a patch.
Thanks for reporting this

**Reply by p-antonioni** [reply]
Thank you for your time and support!

**Reply by GiudGiud** [reply]
Patch PR https://github.com/idaholab/moose/pull/27050

**Comment by ngrilli** [comment]
@p-antonioni  I am not aware of the need to use 6 columns,
if you want to read the Euler angles by grains, meaning each row of the Euler angles file represent one grain,
then you only need three columns with the Euler angles, you can see a test case here:
https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/neper/polycrystalneper.i

I am not familiar with the option of inputting the voronoi tessellation positions,
I would normally create the polycrystal with neper and then input the ready mesh as described above.

**Reply by p-antonioni** [reply]
Thanks Nicolò for your answer.
I saw this procdure in the PropertyReadFile documentation which states: 
"In grain/voronoi mode, a Voronoi tessellation with "nvoronoi" random centers is either created randomly or read from the first columns of the CSV file, depending on the "use_random_voronoi". The file read should still contain ["nprop"] columns, even though 1-3 columns may be used for the Voronoi tessellation positions."
Thanks a lot for sharing your examples codes which work and were really useful to understand the procedure you use.

Best regards,
Paolo

**Reply by ghost** [reply]
> @p-antonioni I am not aware of the need to use 6 columns, if you want to read the Euler angles by grains, meaning each row of the Euler angles file represent one grain, then you only need three columns with the Euler angles, you can see a test case here: https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/neper/polycrystalneper.i
> 
> I am not familiar with the option of inputting the voronoi tessellation positions, I would normally create the polycrystal with neper and then input the ready mesh as described above.

Hi @ngrilli I followed your instructions for neper mesh with ComputeElasticityTensorCPGrain and GrainPropertyReadFile but when I try to output the elasticity tensor components all of the grains of the polycrystal show the same value though they do show different values for the euler angles. Does this show that the rotation wasn't taken into consideration for the calculation or is it just an output error?thanks!
